opa/pkg/main.go

```diff
@@ -0,0 +1,324 @@
+	}
+	dynamicInterface, err := dynamic.NewForConfig(config)
+	if err != nil {
+		panic(err)
+	}
+	kube, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		panic(err)
+	}
+	groupResources, err := restmapper.GetAPIGroupResources(kube.Discovery())
+	if err != nil {
+		panic(err)
+	}
+	restMapper := restmapper.NewDiscoveryRESTMapper(groupResources)
+	// TODO filter by namespace
+	checkInstances, err := dynamicInterface.Resource(instanceGvr).Namespace("").List(ctx, metav1.ListOptions{})
+
+	if err != nil {
+		panic(err)
+	}
+	actionItems, err := processAllChecks(ctx, checkInstances.Items, restMapper, dynamicInterface)
+
+	if err != nil {
+		panic(err)
+	}
+	outputFormat := Output{ActionItems: actionItems}
+	value, err := json.Marshal(outputFormat)
+	if err != nil {
+		panic(err)
+	}
+	err = ioutil.WriteFile(outputFile, value, 0644)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func processAllChecks(ctx context.Context, checkInstances []unstructured.Unstructured, restMapper meta.RESTMapper, dynamicInterface dynamic.Interface) ([]ActionItem, error) {
+
+	actionItems := make([]ActionItem, 0)
+
+	for _, checkInstance := range checkInstances {
+		var checkInstanceObject customCheckInstance
+		err := runtime.DefaultUnstructuredConverter.FromUnstructured(checkInstance.Object, &checkInstanceObject)
+		if err != nil {
+			return nil, err
+		}
+		logrus.Infof("Starting to process check: %s", checkInstanceObject.Name)
+		check, err := dynamicInterface.Resource(checkGvr).Namespace(checkInstanceObject.Namespace).Get(ctx, checkInstanceObject.Spec.CustomCheckName, metav1.GetOptions{})
+		if err != nil {
+			actionItems = append(actionItems, ActionItem{
+				Title:       checkInstanceObject.Name,
+				Description: "An error occured retrieving the Custom Check for this instance",
+				Remediation: "Make sure that the custom check exists and it is in the same namespace as this instance.",
+				Severity:    0.4,
+				Category:    "Reliability",
+			})
+			logrus.Warnf("Error while retrieving check %+v", err)
+			continue
+		}
+		var checkObject customCheck
+		err = runtime.DefaultUnstructuredConverter.FromUnstructured(check.Object, &checkObject)
+		if err != nil {
+			return nil, err
+		}
+
+		newItems, err := processCheck(ctx, checkObject, checkInstanceObject, restMapper, dynamicInterface)
+		if err != nil {
+			return nil, err
+		}
+		actionItems = append(actionItems, newItems...)
+	}
+	return actionItems, nil
+}
+
+func getKubernetesDataFunction(ctx context.Context, check customCheck, restMapper meta.RESTMapper, dynamicInterface dynamic.Interface) func(rego.BuiltinContext, *ast.Term) (*ast.Term, error) {
+	return func(_ rego.BuiltinContext, a *ast.Term) (*ast.Term, error) {
+
+		str, ok := a.Value.(ast.String)
+		if !ok {
+			return nil, nil
+		}
+		strValue := str.String()
+		if len(strValue) > 0 && strValue[0] == '"' {
+			strValue = strValue[1:]
+		}
+		if len(strValue) > 0 && strValue[len(strValue)-1] == '"' {
+			strValue = strValue[:len(strValue)-1]
+		}
```

Comment:
why the need to strip `"`?

`strings.Trim(strValue, "\"")` is probably the way to go though (unless internal `"` is somehow OK)

Reply:
Unfortunately the Rego library returns the string with quotation marks around it. It's super annoying to me. I was trying to find a `Trim` function and couldn't which is why I have this mess. Thanks for this!